cpp/include/cudf/scalar/scalar.hpp

```diff
@@ -0,0 +1,182 @@
+
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_scalar.hpp>
+#include <rmm/thrust_rmm_allocator.h>
+#include <rmm/mr/device_memory_resource.hpp>
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+// Forward declarations
+template <typename T>
+class numeric_scalar_device_view;
+
+class string_scalar_device_view;
+
+template <typename T>
+class timestamp_scalar_device_view;
+
+namespace cudf {
+
+class scalar {
+ public:
+  scalar() = default;
+  ~scalar() = default;
+  scalar& operator=(scalar const& other) = delete;
+  scalar& operator=(scalar&& other) = delete;
+  
+  scalar(data_type type, bool is_valid = false, cudaStream_t stream_ = 0,
+      rmm::mr::device_memory_resource *mr_ = rmm::mr::get_default_resource())
+   : _type(type), _is_valid(is_valid, stream_, mr_)
+  {}
+
+  /**---------------------------------------------------------------------------*
+   * @brief Construct a new scalar by deep copying the contents of `other`.
+   *
+   * All device memory allocation and copying is done using the
+   * `device_memory_resource` and `stream` from `other`.
+   *
+   * @param other The scalar to copy
+   *---------------------------------------------------------------------------**/
+  scalar(scalar const& other);
+
+  /**---------------------------------------------------------------------------*
+   * @brief Move the contents from `other` to create a new scalar.
+   *
+   * After the move, `other.type() = {EMPTY}`
+   *
+   * @param other The scalar whose contents will be moved into the new scalar
+   *---------------------------------------------------------------------------**/
+  scalar(scalar&& other);
+
+  /**---------------------------------------------------------------------------*
+   * @brief Returns the scalar's logical element type
+   *---------------------------------------------------------------------------**/
+  data_type type() const noexcept { return _type; }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Sets this scalar to null
+   *---------------------------------------------------------------------------**/
+  void set_null() { _is_valid.set_value(false); }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Sets this scalar to valid
+   *---------------------------------------------------------------------------**/
+  void set_valid() { _is_valid.set_value(true); }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Indicates whether the scalar contains a valid value
+   *
+   * @note Using the value when `is_valid() == false` is undefined behaviour
+   * 
+   * @return true Value is valid
+   * @return false Value is invalid/null
+   *---------------------------------------------------------------------------**/
+  bool is_valid() const { return _is_valid.value(); }
+
+  bool* valid_mask() { return _is_valid.get(); }
+
```

Comment:
This is more of a question than a review.

In cudf::column, there are operators to implicitly convert column to column_view & mutable_column_view
and view() and mutable_view() functions to get view objects, but they don't exist here (and it seems like there is only scalar_device_view but no scalar_view).

And how should I pass cudf::scalar as a function input argument? cudf::scalar const&?

Reply:
`column_view` was needed for zero-copy slicing. Since that's not required here, scalar is much simpler and it should simply be passed by (const/) reference. Only when it's passed to a device function, you need to create a view beforehand.